### ISSUE TITLE
sql: a logictest with fakedist config hangs

### ISSUE THREAD

**Issue by yuzefovich** [cockroachdb/cockroach#76218]
I stumbled upon a case when a logic test apparently hangs forever during the cleanup part in `fakedist` config (i.e. all queries in the file succeed, but the test doesn't finish).
```
# LogicTest: fakedist

statement ok
SET vectorize = experimental_always;
CREATE TABLE mixed_type_a (a INT8, b TIMESTAMPTZ, FAMILY (a) , FAMILY(b));
CREATE TABLE mixed_type_b (a INT8, b INTERVAL, c TIMESTAMP, FAMILY (a, b, c));

statement error .* dates and timestamp\(tz\) not supported in mixed-type expressions in the vectorized engine
SELECT * FROM mixed_type_a AS a INNER MERGE JOIN mixed_type_b AS b ON a.a = b.a AND a.b < (now() - b.b)
```
I bisected it to b4e85a5a3aa667d3ec76f1bb5e709dd8a498d82c. Interestingly, if the tables only have single column families, then the test succeeds.

I don't think that any features used in this test are to blame, and there likely exist other repros.

My guess is that the goroutine that blocks the test to cleanup is
```
goroutine 3762 [select]:
github.com/cockroachdb/cockroach/pkg/util/retry.(*Retry).Next(0xc00389e970)
	/home/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/util/retry/retry.go:127 +0x14a
github.com/cockroachdb/cockroach/pkg/sql/catalog/lease.(*Manager).WaitForOneVersion(0xc00037a6e0, {0x4d31b18, 0xc0061f9dd0}, 0x38, {0x4c4b40, 0x3b9aca00, 0x3ff8000000000000, 0x0, 0x0, 0x0})
	/home/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/catalog/lease/lease.go:123 +0x208
github.com/cockroachdb/cockroach/pkg/sql.WaitToUpdateLeases({0x4d31b18, 0xc0061f9dd0}, 0x16ee669, 0xeb2)
	/home/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/schema_changer.go:1026 +0x139
github.com/cockroachdb/cockroach/pkg/sql.(*SchemaChanger).exec.func1(0x0)
	/home/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/schema_changer.go:601 +0x56
github.com/cockroachdb/cockroach/pkg/sql.(*SchemaChanger).exec(0xc0003e50a0, {0x4d31b18, 0xc006407050})
	/home/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/schema_changer.go:684 +0x836
github.com/cockroachdb/cockroach/pkg/sql.schemaChangeResumer.Resume.func1(0x38, 0x0, 0x36)
	/home/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/schema_changer.go:2081 +0x4c8
github.com/cockroachdb/cockroach/pkg/sql.schemaChangeResumer.Resume({0xc006197678}, {0x4d31b18, 0xc006407050}, {0x3a7c7c0, 0xc0038da0b0})
	/home/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/schema_changer.go:2147 +0x3d5
github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).stepThroughStateMachine.func2(0x4d4d6b8, {0x4cd8928, 0xc0038da090}, {0x4d31b18, 0xc006407050}, {0x3a7c7c0, 0xc0038da0b0}, 0xc00389f948)
	/home/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/jobs/registry.go:1123 +0xba
github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).stepThroughStateMachine(0xc0002f3b80, {0x4d4d6b8, 0xc006df36a0}, {0x3a7c7c0, 0xc0038da0b0}, {0x4cd8928, 0xc0038da090}, 0xc001b3e180, {0xc006dde1b8, 0x7}, ...)
	/home/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/jobs/registry.go:1124 +0x6d7
github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).runJob(0xc0002f3b80, {0x4d31a70, 0xc006de0600}, {0x4cd8928, 0xc0038da090}, 0xc001b3e180, {0xc006dde1b8, 0x7}, {0xc0046741f8, 0x16})
	/home/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/jobs/adopt.go:401 +0x54f
github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).resumeJob.func1({0x4d4d6b8, 0xc006df27a0})
	/home/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/jobs/adopt.go:323 +0xfa
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2()
	/home/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:488 +0x12a
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx
	/home/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:478 +0x445
```
but I'm not certain.

Jira issue: CRDB-13030

**Comment by ajwerner** [comment]
Seems like we're leaking leases. It could be that it has to do with restarts in the txn loop. We should probably just release the lease as soon as we acquire it.

**Comment by postamar** [comment]
This all seems to have been fixed by 8ba916b2e4c6c20977baf115dbbba43957ccae0f part of @RaduBerinde 's very welcome effort in replacing these annoying table descriptors entirely with `IndexFetchSpecs`. Closing.
